### PR TITLE
Fix a crash in compiling pattern switch when certain combinations of …

### PR DESCRIPTION
**Pull request by gafter**
…type patterns appear together

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems?id=273713

@dotnet/roslyn-compiler @AlekseyTs @agocke Please review this fix for a RC compiler crash.
@MattGertz I propose to fix this compiler crash in RC.

### CODE REVIEW

File: Roslyn.sln [hunk 1]
```diff
@@ -2658,26 +2658,20 @@ Global
 		{1553DE60-A2B0-4FAF-B1B8-C0A7313781CC}.Release|x64.Build.0 = Release|Any CPU
 		{1553DE60-A2B0-4FAF-B1B8-C0A7313781CC}.Release|x86.ActiveCfg = Release|Any CPU
 		{1553DE60-A2B0-4FAF-B1B8-C0A7313781CC}.Release|x86.Build.0 = Release|Any CPU
-		{6DA08F12-32F2-4DD9-BBAD-982EB71A2C9B}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (agocke) [inline Roslyn.sln]: I wouldn't expect any changes to this file.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs [hunk 2]
```diff
@@ -1331,5 +1331,75 @@ public static bool IsB(char value)
...
+        }
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(273713, "https://devdiv.visualstudio.com/DevDiv/_workitems?id=273713")]
+        public void ExplicitTupleConversion_Crash_2()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: Are tuples involved in this test?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs [hunk 3]
```diff
@@ -1331,5 +1331,75 @@ public static bool IsB(char value)
...
+{
+    public struct ValueTuple<T1, T2>
+    {
+        public T1 Item1;
+
+        public ValueTuple(T1 item1, T2 item2)
+        {
+            this.Item1 = item1;
+        }
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: Please add a field for `Item2` to avoid the appearance that the test depends on an invalid tuple type.

**Comment by MattGertz** [comment]
@Kevin Pilch-Bisson<mailto:kevinpi@microsoft.com>@manishj

**Comment by Pilchie** [comment]
Please respond to PR feedback, and then I can approve.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs [hunk 4]
```diff
@@ -1331,5 +1331,75 @@ public static bool IsB(char value)
...
+}
+class Program
+{
+    public static void Main(string[] args)
+    {
+        object o = args;
+        switch (o)
+        {
+            case System.ValueTuple<int, int> z1:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: >System.ValueTuple<int, int> [](start = 17, length = 27)

Please add a test with tuple type syntax.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs [hunk 5]
```diff
@@ -1331,5 +1331,75 @@ public static bool IsB(char value)
...
+                break;
+            case System.ValueTuple<long, long> z3:
+                break;
+        }
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: >VerifyDiagnostics [](start = 24, length = 17)

Consider testing the behavior of the switch as well.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: Technically, no, but it is the same underlying issue (a kind of conversion that wasn't explicitly handled).

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: That isn't possible, as the tuple type syntax isn't valid for a pattern.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: In that case, the `ValueTuple<,>` declaration can be removed and the test name should be changed.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: Does that mean that we have a test for this scenario elsewhere?

Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: It doesn't mean anything related to testing syntax errors for tuple types. That is outside the scope of the issue being addressed here.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: I think it is in scope of the issue being addressed here. Unless we have an existing test that tries to bind a switch statement that syntactically uses tuple types (and I mean doesn't just stop at parsing), please add one because we would want to ensure it doesn't crash too.

Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: OK, sure.

**Comment by gafter** [comment]
All review comments have been addressed.

**Comment by gafter** [comment]
FYI - I rebased this to eliminate the change to Roslyn.sln.

**Comment by Pilchie** [comment]
Approved pending signoffs and test completion.

**Comment by cston** [comment]
LGTM

**Comment by AlekseyTs** [comment]
LGTM
